Feat/toggle checklist groups

rachel-fischoff opened the pull request:
closes #940 

## Summary 
Creates a Toggle Button for each section of the Checklist (toggle content, toggle keyboard, etc) that allows you to mass open or close all of the details for each section. Additoinally, a toggle all button is created to expand and collapse all the checklist group items on the entire Checklist page. 

## Images

Below is the Toggle Content before being clicked
<img width="1172" alt="Screen Shot 2021-05-21 at 4 58 30 PM" src="https://user-images.githubusercontent.com/59700665/119197960-96822f00-ba56-11eb-83af-de43b7397f19.png">

Below is the Toggle Content after being clicked and expanded 
![Screen Shot 2021-05-21 at 4 59 01 PM](https://user-images.githubusercontent.com/59700665/119197983-9f730080-ba56-11eb-8ef0-139a81fd3aaf.png)

Below is the Toggle All button located above the first checklist group
<img width="1214" alt="Screen Shot 2021-05-21 at 4 58 24 PM" src="https://user-images.githubusercontent.com/59700665/119198123-d1846280-ba56-11eb-9d87-3de698558015.png">

## Next Steps
Would love to get some input on the styling especially for the Toggle All button!

ericwbailey reviewed (approved):
This looks good to me! Thank you for submitting it, @rachel-fischoff.

@SaptakS: Could I get a second set of eyes on the JS?

mxmason reviewed (changes requested):
Thanks for writing this PR, @rachel-fischoff, and thank you for your patience with our reviews! Your code looks great! I would like to take care of some things that were not in #940, but which I consider to be blocking.

## Issues
1. These new toggle buttons are not very helpful without JavaScript, so IMO they should be hidden initially, and then shown once we can confirm that JavaScript is on the page.
2. These new buttons need styling in general, so they look like they fit in with our existing design and layout. 
3. IMO, if the checklist items in a category are not all in the same open state, we should either _open all of them_ or _close all of them_. 

### Expanding (ha) issue number 3:

Say the user opens "Use plain language..." to expand it so only that item is open while all the others are closed. Then the user clicks "Toggle Content".
- **Current behavior:** All of the items in the category flip states. "Use plain language..." is closed; all the others open
-  **Desired behavior:** All of the items in the category are opened or closed.


## Next steps
- [ ]  Decide how we ship this. Do we improve the style of the buttons, ship the feature, then come back and improve the open/close state management, or do we do all of this in one PR?
- [ ] Decide what to do in the case of mixed states.
 - [ ] In the case of mixed states, like above, what should the next state be – open, or closed?
 - [ ] How will we handle the `Toggle all` button if one or more items in any category is open?

mxmason commented on src/js/checklist.js:
```diff
@@ -58,10 +58,44 @@ function populateChecklistFromLocalStorage(checkboxSelector) {
+
+renderToggleAll();
+
+function renderToggleCategories(){
```
🔍 You can probably write a query selector that gets both `#toggle-all` _and_ all the other buttons, which then means you won't need a dedicated function for `renderToggleAll`.

rachel-fischoff replied:
Done!

mxmason commented on src/js/checklist.js:
```diff
@@ -58,10 +58,44 @@ function populateChecklistFromLocalStorage(checkboxSelector) {
+function registerToggleButton(buttonEl, parentEl) {
+	if (buttonEl && parentEl) {
+		var details = parentEl.querySelectorAll("details");
+		buttonEl.addEventListener("click", function (event) {
```
🔍 Can we store this event handler as a named function, so it's easier for us to find later when we need it?

rachel-fischoff replied:
yes! Doing it currently! I was going to change the named function of the event listener to changeAttribute. Does that name work? @mxmason

rachel-fischoff replied:
Done! Is handleToggleButtonClick a good name for it?

mxmason reviewed (changes requested):
Thanks for writing this PR, @rachel-fischoff, and thank you for your patience with our reviews! Your code looks great! I would like to take care of some things that were not in #940, but which I consider to be blocking.

## Issues
1. These new toggle buttons are not very helpful without JavaScript, so IMO they should be hidden initially, and then shown once we can confirm that JavaScript is on the page.
2. These new buttons need styling in general, so they look like they fit in with our existing design and layout. 
3. IMO, if the checklist items in a category are not all in the same open state, we should either _open all of them_ or _close all of them_. 

### Expanding (ha) issue number 3:

Say the user opens "Use plain language..." to expand it so only that item is open while all the others are closed. Then the user clicks "Toggle Content".
- **Current behavior:** All of the items in the category flip states. "Use plain language..." is closed; all the others open
-  **Desired behavior:** All of the items in the category are opened or closed.


## Next steps
- [x]  Decide how we ship this. Do we improve the style of the buttons, ship the feature, then come back and improve the open/close state management, or do we do all of this in one PR?
- [x] Decide what to do in the case of mixed states.
  - [x] In the case of mixed states, like above, what should the next state be – open, or closed?
  - [ ] How will we handle the `Toggle all` button if one or more items in any category is open?
  
Thoughts, @ericwbailey and @rachel-fischoff?

ericwbailey commented:
Hi @rachel-fischoff!

First off I want to apologize for the delay in responding to this. I've been going through a ton of work-related things and it's taken a lot of time I'd normally have for work here.

This is going to be a very welcome addition to the site, and I'm really happy you did the work. For my perspective on @mxmason's questions:

>Decide how we ship this. Do we improve the style of the buttons, ship the feature, then come back and improve the open/close state management, or do we do all of this in one PR?

For styling, I'm very familiar with this site's Sass, so I was figuring I could create a followup styling PR once the logic is done/

As for the logic, I think we should get it into the desired state in this PR.

>Decide what to do in the case of mixed states.

My thinking here is:

Toggle all:

- If pressed, all checklist items convert to open state, regardless of their category or prior state.
- If pressed again, checklist items revert to a closed state.

Toggle section:

- If pressed, all checklist items in the relevant section convert to open state, regardless of their category or prior state.
- If pressed again, checklist items update to a closed state.

I am modeling my thinking here after similar OS UI behavior in terms of external consistency (I'm thinking file trees and similar kinds of interactive nested list content). 

The other thing that I think is tangentially relevant is [toggle states and accessible names](https://sarahmhigley.com/writing/playing-with-state/). I don't think this is relevant here directly, in that the existing accessible name plus the ARIA we're using should be sufficient?

rachel-fischoff commented:
Thank you @mxmason & @ericwbailey for this very thorough feedback! I will get to work on the changes early next week! I'll check in if I have any additional questions!

ericwbailey commented:
Yes! Really looking forward to it 🙂

rachel-fischoff commented on src/checklist.njk:
```diff
@@ -93,6 +95,7 @@ templateClass: template-checklist
 	<p class="c-preface">
 		{{ content.preface | safe }}
 	</p>
+	<button data-toggle-category="{{ category }}">Toggle {{ category }}</button>
```
I'm happy to add a class. As a new(er) dev, was just curious about the reasoning.

rachel-fischoff replied:
Added a class!

rachel-fischoff commented:
I know this is almost 2 years old but should I close this PR @mxmason @ericwbailey or should I still try to get this functionality merged in? Not sure if it's even desired anymore.

SaptakS commented:
I think @ericwbailey might be able to comment better, but I would definitely love to see this PR merged given you still have the capacity. I can do a re-review to see if all the above concerns are addressed and what is left if that helps. I think @ericwbailey said he can make a commit for the  design separately once functionality is done. In case he is not available, I can use existing designs in the sass code to make it look similar to the rest of the website.

rachel-fischoff commented:
@SaptakS I think I had addressed all the comments beforehand and was hading to hear back. If I recall, I was just waiting on the design but let me know about functionality.

SaptakS commented:
I will take a look.

SaptakS commented on src/js/checklist.js:
```diff
@@ -58,10 +58,45 @@ function populateChecklistFromLocalStorage(checkboxSelector) {
+		buttonEl.addEventListener("click", function handleToggleButtonClick(event) {
+			var target = event.target;
+			if (target.hasAttribute("data-open") === true) {
+				target.removeAttribute("data-open");
```
Can we instead use `aria-expanded="true"` in place of data-open? I think that will help convey meaningful information to the assistive technology about the state of the button as well, and helps with the logic as well.
For reference: https://adrianroselli.com/2020/05/disclosure-widgets.html#Script

SaptakS commented on src/js/checklist.js:
```diff
@@ -58,10 +58,45 @@ function populateChecklistFromLocalStorage(checkboxSelector) {
+
+function registerToggleButton(buttonEl, parentEl) {
+	if (buttonEl && parentEl) {
+		var details = parentEl.querySelectorAll("details");
```
Can we make this selector a little more specific? Like `details.c-checklist` to ensure that we are not toggle any other `<details>` that might be there. Just to future-proof it a bit.

SaptakS reviewed (changes requested):
The code seems to be not functioning right now. I left a few comments which I think should help resolve that. Also, @mxmason mentioned that we should hide the toggle buttons by default and only show them from the javascript code, because the toggle buttons don't work without javascript, and hence shouldn't be visible for non-js users.

rachel-fischoff commented:
@SaptakS will address this weekend and get back to you!

rachel-fischoff commented:
@SaptakS Sorry, I've been sick but I'll message you when I have the fix.

SaptakS commented:
@rachel-fischoff no need of apologies and no hurries. Take your time! And I hope you feel better soon.

rachel-fischoff commented:
Hi @SaptakS and team. I keep overestimating my capacity and haven't been in JavaScript land in a long time. I think it makes the most sense to just close this PR and then comment on the issue that I never got around to it. If someone wants to open it back up or continue this work then they are more than welcome.

SaptakS commented:
That's totally fine! I don't think this PR needs to be closed. Especially since it doesn't have any merge conflicts. I think anyone (or one of us) who wants to work on the issue can continue on this PR. I haven't checked however if this branch can be pushed into by us. If not, I might create a new PR with your commits and continue from there. But I don't want to close this PR and make your work go away. :) Thanks for your contributions!

ashimupd commented:
@ericwbailey @SaptakS @rachel-fischoff @mxmason Hi guys, hope you are doing well, I have had this PR pending for a few weeks: https://github.com/a11yproject/a11yproject.com/pull/1612 related to this.
